### DISCUSSION TITLE
Optimal hardware setup (GPUs)

### DISCUSSION THREAD

**Discussion by SuperKogito** [coqui-ai/STT#2296] [General]
I am trying to speed up my training process. Current setup: 1 GPU RTX 2080 8Gb (12 CPUs server + 32 gb ram) and I am training using over 1k hours of data. Unfortunately, the process is very slow; it takes almost 1 day for only 1 epoch and I can only use a batch size of 1 (over 4Gb are used in the GPU ram). My questions here are: 
- Why does the training occupy so much space in the GPU ram; isn't it supposed to load the model weights and the voice file and that's it? shouldn't I be able to use larger batches? 
- What hardware / setup can you recommend to speed the process? (I have seen people saying that 2 GPUs will help but not necessarily 4).
- Any other recommendations on how to speed things up ?  

I appreciate all the advice and details of your experience. 
Thank you in advance :))

**Comment by wasertech** [comment]
> I am trying to speed up my training process.

You need more VRAM per GPU to fit bigger batch sizes.
> I have seen people saying that 2 GPUs will help

Adding another 8 Gb GPU won't really help.
> Why does the training occupy so much space in the GPU ram; isn't it supposed to load the model weights and the voice file and that's it?

You need to fit 16 or 32 batched samples 3 times on each GPU so try to aim for 24 Gb for the biggest batch sizes (64, 128) but 12 Gb or 11 Gb might just cut it for you.
> What hardware / setup can you recommend to speed the process?

Only then you can think about adding more GPUs of the same size to make training even faster.
> shouldn't I be able to use larger batches?

Try to limit the length of each sample (10s < 20s). This is the biggest factor in the batch size computation so use it carefully.
> Any other recommendations on how to speed things up ?

Use automatic mixed precision ([AMP](https://stt.readthedocs.io/en/latest/MIXED_PRECISION.html?highlight=automatic%20mixed)) to enable bigger batch sizes and make the training faster but use it only for exploring with different parameters. Once you are happy with your tests, you can make a final training without AMP to export your final model.

**Reply by SuperKogito** [reply]
@wasertech  thank you a lot for your response. It sure answers a lot of my questions. If I may ask more:
- Why do I need to fit batched samples 3 times, shouldn't this be 1time?
- also any idea; what is exactly occupying the GPU vram? it cannot be just the model and audio that is reaching over 4gb.
- The requirement of 24 Gb of VRAM narrows down my choices quite a bit, is there any GPUs/ rig setup do you recommend?

**Reply by wasertech** [reply]
> Why do I need to fit batched samples 3 times, shouldn't this be 1time?

https://github.com/coqui-ai/STT/blob/fcec06bdd89f6ae68e2599495e8471da5e5ba45e/training/coqui_stt_training/train.py#L326-L340

It's the standard batch size test. If you can't pass it without crashing, it's more than likely that you can't train with your current batch size.

> what is exactly occupying the GPU vram?

```python
use_allow_growth: bool = field(
        default=False,
        metadata=dict(
            help="use Allow Growth flag which will allocate only required amount of GPU memory and prevent full allocation of available GPU memory"
        ),
    )
```
You can allow memory growth ([docs](https://stt.readthedocs.io/en/latest/TRAINING_FLAGS.html?highlight=allow%20growth)) to see exactly what's used but it's not recommended unless you really need it because it can cause out-of-memory issues if you are not careful with your memory consumption.
> The requirement of 24 Gb of VRAM narrows down my choices quite a bit, is there any GPUs/ rig setup do you recommend?

It's a target not a requirement, if it's more convenient for you to use batch size 32 on 4x 12 Gb GPUs rather than batch size 64 on 2x 24 Gb GPUs, then go for it.

As for a rig recommendation, these tend to become obsolete quite fast so I won't really indulge you. 
What I can say is that:
- a good threadripper helps with data manipulation,
- enough fast solid memory to store this data close the CPU is a must,
- enough speedy RAM for the CPU to do the job,
- the largest VRAM per GPU you can get on board, (i.e. mine is 2x 24 Gb/GPU)
- A beefy power supply or it will trigger current protections during training.

**Reply by wasertech** [reply]
> It's a target not a requirement, if it's more convenient for you to use batch size 32 on 4x 12 Gb GPUs rather than batch size 64 on 2x 24 Gb GPUs, then go for it.

I'm not saying you should go buy 4x 12Gb GPU right now! Add one to begin with. Even with batch size 32 you'll be able to train in a more reasonable time than with batch size 1. :smile: 

You can always add more later.

**Reply by HarikalarKutusu** [reply]
I would add good cooling for the PC and some air conditioner for the room + extra funds to pay the increased electrical bill...
Or go for Google Colab Pro like I do, you can even work from the beach :) (joke)

**Reply by SuperKogito** [reply]
@HarikalarKutusu thank you for weighing in on this. I was actually considering an online service like AWS or Google, but isn't it cheaper / more affordable to buy your own hardware and not to rent it?

**Reply by HarikalarKutusu** [reply]
It really depends on what you are doing.
- If you run trainings continuously and want to leave it overnight(s) for batch processing, have big datasets, e.g. it is your profession, it IS better to have the hardware. **Cons**:  Investment money, electrical bill, location dependency, upgrade costs (new HW comes out), your GPU will not be better than cloud ones. I must add clean energy to this, in here (Turkey) the energy production is not the cleanest.
- If you're experimenting like me, training occasionally and/or want to change parameters interactively cloud sources might be better. AWS or Google Pro+ are costly for this purposes, so I went for $10/month Google Plus for interactive work on notebooks. **Pros**: I get P100 16GB PCI, which runs on more clean energy, Dl speeds are high (20-80 MBps) wrt my residential connection (100/5 Mbps), and if you have a good amount of Google Drive space, you keep your data there... **Cons**: Not good for large datasets, you have max 24 hours for a cell run. I measured I can work with 100-150 hours of data with that. Not good for batch processing, you need to be online on your end. Notebooks are good for experimenting & educational purposes but not for production work. There are limits, you should not abuse. Download/upload of datasets from your machine sucks.

If you want batch processing, AWS is better of course, but more costly (not tried).

Last year I checked my options and decided to go this way. There also was that chip shortage and incredible GPU prices.

Now, I'm in my summer house on south, at the balcony, training on multiple Common Voice datasets (you can have 4 non-GPU tasks such as LM generation + 1 GPU task such as training).

**Reply by wasertech** [reply]
> I would add good cooling for the PC and some air conditioner for the room + extra funds to pay the increased electrical bill

It's all in the IT budget, yes. Cooling is really a science in itself but you need good air/liquid flow to keep your silicon cool. Yes, you need energy to compute but it's cheaper in the long term to own your hardware, specially if you use it a lot over a long period of time.
> [Hardware] Cons: 
> - Investment money, 
> - electrical bill, 
> - upgrade costs (new HW comes out), 

Initial investment and upkeep cost is not a con. You'll end up paying more if you borrow your hardware and that's factoring cost of energy.
I might also add that we are not mining on a blockchain here. It's punctual use of the hardware meaning you don't spend a lot of energy all the time. Just when you need it.
> - location dependency, 

It's a dumb feature. Why would I want to tie my GPUs to the internet?
If I really wanted my GPU to be accessible over the network I could rent 2U and send my rack to my local services provider for a reasonable price. And it's still my computer!
> - your GPU will not be better than cloud ones.
> I get P100 16GB PCI

So you get 1x 16Gb GPU? I already have 2x 24Gb GPUs... In which way colab's config is better?
Even if it where the case, I'd rather fully own my compute medium rather that to borrow half a computer that someone else can spare but that's just me I guess.
> I must add clean energy to this, in here (Turkey) the energy production is not the cleanest.

I don't live in Türkiye. My local energy is mostly renewable (my hosting platform even manages 100% clean electricity without using air conditioning to cool their data-centers).

Each time I do the math I come to this conclusion:
> [About Colab limitations] (you can have 4 non-GPU tasks such as LM generation + 1 GPU task such as training).
> There are limits, you should not abuse. Download/upload of datasets from your machine sucks.
> and if you have a good amount of Google Drive space

So cute. I have an old theadripper with only 12 cores (meaning up to 24 simultaneous tasks), two fast and spacious GPUs for my computing needs plus 12 Tb of space to store it all. I don't need an internet connection to use it and I can keep it all forever to use as I see fit.

No amount of cloud services can replace your personal computer.

@SuperKogito Wanted the "Optimal hardware setup (GPUs)" for training. Cloud is nowhere optimal IMO specially if you compare it to owning the hardware. I would rather use a cloud provider to serve my models rather than to train them. :man_shrugging:

**Reply by HarikalarKutusu** [reply]
Chill :)

> It really depends on what you are doing.

**Reply by SuperKogito** [reply]
Thank you a lot for your posts <3 You both raise great points and argument two good possible options depending on the situation. 
In my case owning my hardware seems to be a better option since I intend to train permanently and for different languages with over 1k hours if possible. Coqui developers used a server with 8 NVIDIA A100 GPUs each with 40GB of VRAM), but unfortunately I cannot even afford 1 A100 xD instead I am thinking of 2x RTX 3090 (24 Gb VRAM). I would go with 4x but according to different posts it doesn't bring a lot.

> You need to fit 16 or 32 batched samples 3 times on each GPU so try to aim for 24 Gb for the biggest batch sizes (64, 128) but 12 Gb or 11 Gb might just cut it for you.

This still doesn't make sense to me. Assume a wave audio of 20 seconds this is equivalent to 1.28 MB x 3 = 0.00384 GB so based on this logic I should be able to have a batch size of a 1000 (0.00384 *1000 = 3.84 GB)  and I will still have over 4 GB free for the model. That's why I really do not get this VRAM constraint.

**Reply by wasertech** [reply]
There is quite a lot missing in your reasoning.

> Assume a wave audio of 20 seconds this is equivalent to 1.28 MB x 3 = 0.00384 GB so based on this logic I should be able to have a batch size of a 1000 (0.00384 *1000 = 3.84 GB)

For starters, 1.28 Mb times 3 is equal to 0,03072 Gb not 0.00384. So with a 8Gb card, you can only fit 100 batched samples not 1000.

You also need to account for the storage of the weight parameters, the activations from the last forward pass must also be retained until they can be used to calculate the error gradients in the backwards pass. There is also constraints related to the hardware architecture and so much more that comes into account.

To go further, if you’re interested, I suggest [this Graphcore article about NN memory consumption](https://www.graphcore.ai/posts/why-is-so-much-memory-needed-for-deep-neural-networks). It’s not specific to STT but the principles are the same.

**Reply by SuperKogito** [reply]
This a great read @wasertech, Thank you so much (My apologies for the bad math smh).

**Reply by HarikalarKutusu** [reply]
Pardon me asking, but:
- Isn't 1.28 MB = 0.00125 GB (approx) 
- How do you get 1.28 MB for 20 sec WAV file? Do you train on stereo and/or 48 KHz etc?

**Reply by wasertech** [reply]
> Isn't 1.28 MB = 0.00125 GB (approx)

![image](https://user-images.githubusercontent.com/79070834/191130608-d15b0692-de07-4267-bc76-bb52e6be400f.png)
@HarikalarKutusu  I'll go with my calc on this one 😅

> How do you get 1.28 MB for 20 sec WAV file? Do you train on stereo and/or 48 KHz etc?

@SuperKogito Yeah it's quite heavy. As an example, here is a 16 sec converted wav file (pre-processed for STT) from AttHack ([OpenSLR 88](http://www.openslr.org/88/)). 

```zsh
❯ soxi M19_a3_s044_v01_.wav

Input File     : 'M19_a3_s044_v01_.wav'
Channels       : 1
Sample Rate    : 16000
Precision      : 16-bit
Duration       : 00:00:16.83 = 269280 samples ~ 1262.25 CDDA sectors
File Size      : 539k
Bit Rate       : 256k
Sample Encoding: 16-bit Signed Integer PCM

❯ du -h M19_a3_s044_v01_.wav
528K    M19_a3_s044_v01_.wav
❯ ls -l M19_a3_s044_v01_.wav
.rw-rw-r-- root root 526 KB Tue May 17 14:01:41 2022  M19_a3_s044_v01_.wav
```

It weights only 526 KB.

**Reply by HarikalarKutusu** [reply]
> @HarikalarKutusu I'll go with my calc on this one 😅

Please select gigabytes, not Gigabits 🔢, you are multiplying with 8...

@SuperKogito, if you will be using Coqui STT, you need 16 KHz mono WAV files. For those who like calculators, here is one:
https://www.colincrawley.com/audio-file-size-calculator/

The formula is (results is in BYTES):
file_size = sample_rate * (bit_depth / 8) * channels * duration

As it gives  625/640, probably you were checking on a stereo file?

Perhaps use GB instead of Gb on GPU memory sizes etc?

**Reply by SuperKogito** [reply]
I actually used the same calculator but I might have not gave the right inputs. However, it still made my point that the size of the audios is actually small compared to the occupied memory on the VRAM.

**Reply by FivomFive** [reply]
@wasertech it is more convenient for you to use more batches on a single with a lot of memory than multiple gpus. As I understand multi gpu usage correctly one will also get a bigger effective batchsize but with the overhead of aggregating parameter updates. Aggregation step should be much lower than the time spend to evaluate each batch.
I'm using a batchsize of 128 on each gpu which should produce highest compute time but using 4 gpus I get a low gpu utilization and bad scaling as described in #2258 

You seem to be more into it can you explain what is the problem with using multiple gpus instead of one with larger memory?

**Reply by wasertech** [reply]
@FivomFive I don't have 4 GPUs so I can really test anything. My best guess, is that it's expected. Adding more GPU doesn't scale well. At one point it's better to upgrade your GPUs with more memory rather than to add another. You already are reaching the limits of the hardware, so I can't really do much for you. 🤷
